cardinal/config.go

```diff
@@ -33,39 +48,104 @@ var defaultConfig = WorldConfig{
 type RunMode string
 
 type WorldConfig struct {
-	RedisAddress              string  `config:"REDIS_ADDRESS"`
-	RedisPassword             string  `config:"REDIS_PASSWORD"`
-	CardinalNamespace         string  `config:"CARDINAL_NAMESPACE"`
-	CardinalMode              RunMode `config:"CARDINAL_MODE"`
-	BaseShardSequencerAddress string  `config:"BASE_SHARD_SEQUENCER_ADDRESS"`
-	BaseShardQueryAddress     string  `config:"BASE_SHARD_QUERY_ADDRESS"`
-	CardinalLogLevel          string  `config:"CARDINAL_LOG_LEVEL"`
-	StatsdAddress             string  `config:"STATSD_ADDRESS"`
-	TraceAddress              string  `config:"TRACE_ADDRESS"`
+	// Cardinal
+	CardinalMode      RunMode `config:"CARDINAL_MODE"`
+	CardinalNamespace string  `config:"CARDINAL_NAMESPACE"`
+	CardinalLogLevel  string  `config:"CARDINAL_LOG_LEVEL"`
+
+	// Redis
+	RedisAddress  string `config:"REDIS_ADDRESS"`
+	RedisPassword string `config:"REDIS_PASSWORD"`
+
+	// Shard networking
+	BaseShardSequencerAddress string `config:"BASE_SHARD_SEQUENCER_ADDRESS"`
+	BaseShardQueryAddress     string `config:"BASE_SHARD_QUERY_ADDRESS"`
+
+	// Telemetry
+	StatsdAddress string `config:"STATSD_ADDRESS"`
+	TraceAddress  string `config:"TRACE_ADDRESS"`
+}
+
+func LoadWorldConfig() WorldConfig {
+	cfg := defaultConfig
+
+	// Load config from environment variables
+	if err := config.FromEnv().To(&cfg); err != nil {
+		log.Fatal().Err(err).Msg("Failed to load config")
+	}
+
+	// Validate config
+	if err := cfg.Validate(); err != nil {
+		log.Fatal().Err(err).Msg("Invalid config")
+	}
+
+	// Set logger config
+	if err := cfg.setLogger(); err != nil {
+		log.Fatal().Err(err).Msg("Failed to set log level")
+	}
+
+	return cfg
 }
 
-func (w WorldConfig) Validate() error {
-	if w.CardinalMode != RunModeProd {
-		return nil
+func (w *WorldConfig) Validate() error {
+	// Validate Cardinal config
+	if w.CardinalMode != RunModeProd && w.CardinalMode != RunModeDev {
+		return eris.Errorf("CARDINAL_MODE must be either %q or %q", RunModeProd, RunModeDev)
+	}
+	if err := Namespace(w.CardinalNamespace).Validate(w.CardinalMode); err != nil {
+		return eris.Wrap(err, "CARDINAL_NAMESPACE is not a valid namespace")
+	}
+	if w.CardinalLogLevel == "" && slices.Contains(validLogLevels, w.CardinalLogLevel) {
+		return eris.New("CARDINAL_LOG_LEVEL must be one of the following: " + strings.Join(validLogLevels, ", "))
+	}
+
+	// Validate Redis config
+	if _, _, err := net.SplitHostPort(w.RedisAddress); err != nil {
+		return eris.New("REDIS_ADDRESS must follow the format <host>:<port>")
+	}
+	if w.CardinalMode == RunModeProd && w.RedisPassword == "" {
+		return eris.New("REDIS_PASSWORD must be set in production mode")
 	}
-	if w.RedisPassword == "" {
-		return eris.New("REDIS_PASSWORD is required in production")
+
+	// Validate shard networking config
+	if w.CardinalMode == RunModeProd {
```

Comment:
lets group the checks we make for when mode == `RunModeProd` under a single if statement.

Reply:
i considered that, but decided against it because I feel it's more readable to be able to go line by line against the order of the struct property definition.

Reply:
I feel like its more important to be able to look at a block of code and uniformly say "oh these are all the fields that are required by prod mode" 

field ordering never felt that important to me bc i can just go to the fields and press cmd+b and find their usages

Reply:
it's less so about finding the property, but more so to make sure you are not missing validation on a field

Reply:
i still think it would be nice to have all prod mode required variables under one block of code